### PR TITLE
Default to no unit for `dtype=bool`

### LINKED ISSUE #2396: Unit of comparison results
Currently element-wise comparison operators such as `==` or `<` return a variable with unit `dimensionless`. With the introduction of variables without unit (`var.unit == None`) in #2395 the question is whether this is still the right choice. Should we use `None` instead?

On the surface, one argument against the use of `None` would be that we may want to do arithmetic based on the result of a comparison operator. This would be prevented if `None` was use.
Historically this has however led to bugs such as those fixed in, e.g., https://github.com/scipp/scipp/pull/2133 (and if I remember correctly this was not the first such bug). If the bool-valued masks had had `unit == None` then multiplying the mask into the data to "apply" the mask would not have been possible, and the bug (treating NaN incorrectly) would have been prevented. It is likely that users may try the some faulty approach.

Another argument is that `True` or `False` are not quantities (except for very special cases when they may be used for a very small problem domain to save memory), i.e., saying "`True` is a dimensionless quantity" sounds quite odd.

Even just considering the implementation, using `None` would make more sense. For example, say we compare `5.0 kg == 6.0 kg`. `kg == kg` may be `True` but not `dimensionless`. That is, the result is not in the space of possible units. What justifies using `dimensionless`?

A related question is that of the (default) unit used for integers. Obviously `5 kg` *is* a sensible quantity, but the use of integers is a special case and in most cases integers appear to be used as "indices". Indices are not quantities. Similar to the bugs from arithmetic with bool-valued variables, we can imagine errors from using indices in arithmetic: `da.coords['detector_id'] * da.data` is currently valid, since the `detector_id` coord is a "dimensionless quantity". If we were to default to using `unit == None` for integer-valued variables then this operation (which is not sensible, since `detector_id` is an arbitrary ID with no physical meaning) would be prevented.

The above discussion would imply:

- Change default-unit of integer dtypes (including `bool`) to `None`. Note that without this default label-based indexing such as `da['detector_id', sc.scalar(2342)]` would not work (`unit-None` would have to be specified explicitly).
- Change unit returned by element-wise comparison operators to `None`
- Adapt logical operators to handle this.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
- Fixes #2396 by setting default unit of `bool` to `None` and adapting comparison on logical operators.
  - Comparison operators return results with `unit=None`. 
- Change default unit of `index_pair` (and thus indices used for binned data) to `None`.

There is one thing here that feels odd: I had to do some workaround for `mean` which normalizes to the element count. Is there a saner definition of some operation that could avoid this? Should `sum` return `dimensionless` if the input has unit `None`?

### CODE REVIEW

**Comment by jl-wynen** [comment]
> Should sum return dimensionless if the input has unit None?

That sounds strange. `__add__` doesn't do it...

File: docs/user-guide/binned-data/computation.ipynb [hunk 1]
```diff
@@ -214,7 +214,7 @@
    "cell_type": "code",
    "execution_count": null,
    "metadata": {
-    "scrolled": true
+    "tags": []
```
Comment (jl-wynen) [inline docs/user-guide/binned-data/computation.ipynb:217]: Remove?
I had an empty tag appear in a PR, too. Not sure where they are coming from

File: docs/user-guide/coordinate-transformations.ipynb [hunk 2]
```diff
@@ -287,8 +287,9 @@
     "                      'location': location,\n",
     "                      'datetime': datetime\n",
     "                  })\n",
-    "da += 2. * (location.fields.z > 0. * sc.Unit('km')).astype(\n",
-    "    'float64')  # more lightning strikes in northern hemisphere\n",
+    "north = location.fields.z > 0. * sc.Unit('km')\n",
+    "north.unit = ''\n",
```
Comment (jl-wynen) [inline docs/user-guide/coordinate-transformations.ipynb:291]: What does that do? Does it set the unit to 'one'? If so, better spell it out, I think.

File: lib/dataset/bins.cpp [hunk 3]
```diff
@@ -380,16 +380,19 @@ Variable bins_sum(const Variable &data) {
 Variable bins_mean(const Variable &data) {
   if (data.dtype() == dtype<bucket<DataArray>>) {
     const auto &&[indices, dim, buffer] = data.constituents<DataArray>();
-    if (const auto mask_union = irreducible_mask(buffer.masks(), dim);
+    if (auto mask_union = irreducible_mask(buffer.masks(), dim);
         mask_union.is_valid()) {
       // Trick to get the sizes of bins if masks are present - bin the masks
       // using the same dimension & indices as the data, and then sum the
       // inverse of the mask to get the number of unmasked entries.
+      mask_union.setUnit(units::one);
```
Comment (jl-wynen) [inline lib/dataset/bins.cpp:388]: Why is this needed?

File: lib/variable/operations_common.h [hunk 4]
```diff
@@ -54,3 +55,12 @@ template <class T, class Op> auto reduce_all_dims(const T &obj, const Op &op) {
 }
 
 } // namespace scipp::variable
+
+namespace scipp {
+template <class Var, class... Dim>
+auto count_finite(const Var &var, Dim &&... dim) {
+  auto count = sum(isfinite(var), dim...);
+  count.setUnit(units::one);
```
Comment (jl-wynen) [inline lib/variable/operations_common.h:63]: Isn't this conceptually cleaner?
May be inconvenient though, so I am not saying that it should be changed.
```suggestion
  count.setUnit(units::count);
```

File: src/scipp/core/variable.py [hunk 5]
```diff
@@ -55,6 +55,19 @@ def scalar(value: _Any,
                          dtype=dtype)
 
 
+def index(value: _Any, dtype: _cpp.DType = None) -> _cpp.Variable:
+    """Constructs a zero dimensional :class:`Variable` representing an index.
+
+    This is equivalent to calling :py:func:`scipp.scalar` with unit=None.
+
+    :param value: Initial value.
+    :param dtype: Optional, type of underlying data. Default=None,
```
Comment (jl-wynen) [inline src/scipp/core/variable.py:64]: Should we enforce `int32` or `int64`?

File: tests/bins_test.py [hunk 6]
```diff
@@ -186,48 +188,6 @@ def test_lookup_getitem(dtype):
     assert sc.identical(lut[var], expected)
 
 
-def test_load_events_bins():
-    with in_memory_nexus_file_with_event_data() as nexus_file:
```
Comment (jl-wynen) [inline tests/bins_test.py:190]: Huh, didn't realise we still had this in scipp. Good find!

File: docs/user-guide/binned-data/computation.ipynb (hunk 1 above)
Comment (SimonHeybrock) [reply docs/user-guide/binned-data/computation.ipynb:217]: I think this came from previously having a `scrolled: true`, which I removed, and now there is an empty tags instead?

File: lib/dataset/bins.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply lib/dataset/bins.cpp:388]: The unit if the masks is `none`, but we need to normalize to it.

File: lib/variable/operations_common.h (hunk 4 above)
Comment (SimonHeybrock) [reply lib/variable/operations_common.h:63]: This is computing `N` in the denominator of `mean = sum/N`, so `count` is not useful here.

File: src/scipp/core/variable.py (hunk 5 above)
Comment (SimonHeybrock) [reply src/scipp/core/variable.py:64]: I had considered that, but in the end I could not come up with a compelling argument for it, because I could not identify any risk of allowing, say, float.

I think we *do* want to support other types that default to no unit anyway. For example, someone may want to use Python tuples as indices (`sc.index((1,2))`, since `sc.scalar((1,2))` is odd), so unless we want to selectively forbid floating-point dtypes...

File: lib/dataset/bins.cpp (hunk 3 above)
Comment (jl-wynen) [reply lib/dataset/bins.cpp:388]: Is the reason that `normalize_impl` wants to divide by the sum of masks?

File: src/scipp/plotting/controller1d.py [hunk 7]
```diff
@@ -33,6 +33,9 @@ def rescale_to_data(self, button=None):
         with_min_padding = self.vmin is None
         with_max_padding = self.vmax is None
         vmin, vmax = self.find_vmin_vmax(button=button)
+        if vmin.unit is None:
+            vmin.unit = ''
+            vmax.unit = ''
```
Comment (jl-wynen) [inline src/scipp/plotting/controller1d.py:38]:
```suggestion
            vmin.unit = 'one'
            vmax.unit = 'one'
```

In my opinion, we should make this a (informal) style guide. I find it confusing to use both empty strings and `None` even though both mean different things.

File: lib/dataset/bins.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply lib/dataset/bins.cpp:388]: Yes. I am just looking into #2224 to see if there is a better solution.

**Comment by SimonHeybrock** [comment]
@jl-wynen I avoided some of the unit handling for normalization when computing mean now. This is ready for another look.

File: lib/variable/operations_common.h [hunk 8]
```diff
@@ -28,15 +28,22 @@ SCIPP_VARIABLE_EXPORT Variable &nanmean_impl(const Variable &var, const Dim dim,
...
+      numerator.dtype() == dtype<float> ? dtype<float> : dtype<double>;
+  denominator.setUnit(units::one);
+  return numerator *
          reciprocal(astype(denominator, type, CopyPolicy::TryAvoid));
 }
 
+template <class T> void normalize_inplace_impl(T &numerator, T denominator) {
+  denominator.setUnit(units::one);
+  numerator *= reciprocal(
+      astype(denominator, core::dtype<double>, CopyPolicy::TryAvoid));
```
Comment (jl-wynen) [inline lib/variable/operations_common.h:44]: Why do both `normalize_*_impl` functions multiply by the reciprocal instead of dividing by the denominator? That would avoid the intermediate variable.

Comment (SimonHeybrock) [reply lib/variable/operations_common.h:44]: I think the reason was that usually the denominator is small (may just be a scalar), so it is more important to use a cheap `*` instead of `/` for every element.

**Review by jl-wynen — APPROVED** [review]
